### PR TITLE
SUBSCRIBE and UNSUBSCRIBE support multi-topic filters

### PR DESCRIPTION
**Pull request by yinbaoqiang**
Submit for (Issue #141 )

### CODE REVIEW

**Comment by mwallnoefer** [comment]
@ejvr + @KonstantinRitt What do you think about the patch?

File: src/mqtt/qmqtt_client_p.cpp [hunk 1]
```diff
@@ -378,17 +416,29 @@ void QMQTT::ClientPrivate::onNetworkReceived(const QMQTT::Frame& frm)
...
+    case SUBACK: {
         mid = frame.readInt();
-        topic = _midToTopic.take(mid);
-        qos = frame.readChar();
-        handleSuback(topic, qos);
+        QMap<QString,quint8> topicMap = _midToTopic.take(mid);
+        QMapIterator<QString, quint8> i(topicMap);
+        QMap<QString,quint8> resultMap;
+        quint8 tmpqos = 0;
+         while (i.hasNext()) {
```
Comment (mwallnoefer) [inline src/mqtt/qmqtt_client_p.cpp]: wrong `while` indentation

File: src/mqtt/qmqtt_client_p.cpp [hunk 2]
```diff
@@ -259,6 +259,20 @@ quint16 QMQTT::ClientPrivate::sendSubscribe(const QString & topic, const quint8
     _network->sendFrame(frame);
     return mid;
 }
+quint16 QMQTT::ClientPrivate::sendSubscribes(const QMap<QString, quint8> map)
+{
+    quint16 mid = nextmid();
+    Frame frame(SETQOS(SUBSCRIBE, QOS1));
+    frame.writeInt(mid);
+    QMapIterator<QString, quint8> i(map);
+     while (i.hasNext()) {
```
Comment (mwallnoefer) [inline src/mqtt/qmqtt_client_p.cpp]: wrong `while` indentation

File: src/mqtt/qmqtt_client_p.cpp (hunk 1 above)
Comment (ejvr) [reply src/mqtt/qmqtt_client_p.cpp]: The rest of the code uses for-loops for iteration over QT collections. I would do that here as well, for consistency.

**Comment by ejvr** [comment]
I wonder if we really need this feature. The mqtt standard allows a client to send multiple subscriptions in a single message, but it is not obligatory. I guess the overhead of multiple calls to subscribe is not that big. And subscribe is not called very often.

If we decide to implement this feature, I would prefer to clean up the interface a bit: leave the current subscribed signal as it is and add a new one for multiple subscriptions.

**Comment by yinbaoqiang** [comment]
@ejvr This demand is useful.
For example, when you log in to a system in your program, you need to subscribe to multiple topics and then go to the program home page. A single subscription requires more code to subscribe than to multiple topics.

**Comment by ejvr** [comment]
@yinbaoqiang It may be useful, but what I wanted to know is if there is a need to send multiple subscription is a single request. It is easy to create a function that loops over a container of subscription strings and call subscribe for each one of them. But you pull request adds a lot more complexity, which we need only if a single subscription request with multiple subscriptions is needed.

**Comment by yinbaoqiang** [comment]
@ejvr This is a matter of agreement integrity and compatibility. Since the issue was not considered before, it is now forced to increase, the code complexity becomes higher, and it is not elegant to implement. It is acceptable to not support it.

**Pull request closed without merging**